proposals/p5337.md

```diff
@@ -795,16 +790,57 @@ final match_first {
 fn G(U:! Y) {
   // Can assume `U impls J` and `U.(J.A)` is `bool`.
 }
+
+fn H(V:! W) {
+  // Don't know if `V impls Y`, so we don't know
+  // which `impl` is selected. We use a symbolic
+  // witness for `V` and `V.(J.A)` is unknown
+  // (despite that it would be `bool` in either
+  // case).
+}
 ```
 
-> FIXME _**QUESTION**_: We could say that in this situation the value of `A.J`
-> is known to be `bool` if the query matches `[V:! W] V as J`, even though that
-> `impl` may not be used. The logic is that if that impl is not used, the
-> `impl forall [U:! Y] U as J` will be selected instead and it sets `A.J` to the
-> same thing. This is a bit like what we did in
-> [proposal #2868](https://github.com/carbon-language/carbon-lang/pull/2868)
-> which allowed an `impl` to overlap a `final impl` as long as agreed on the
-> overlap.
+The compiler will only use a known matching final `impl`'s associated constants
+if it is the first that _could_ match, as determined by type structure. For
+example:
+
+```carbon
+interface L(T:! type) {
+  let B:! type;
+}
+class C(T:! type) {}
+
+final match_first {
+  // Can use the associated constants from this impl,
+  // since it is the first, so nothing could match
+  // ahead of it.
+  impl forall [U:! Y] C(U) as L(bool) where .B = i32;
+
+  // Can use the associated constants from this impl
+  // only if V can't be `C(U)`, and so it is known
+  // that it won't use the previous impl instead.
```

Comment:
IIUC this is only known for concrete queries then? As in - any generic in the query self type for `L(bool)` will not receive back a concrete result even though it's from a `final match_first` because there's more than one impl of that interface in the match_first and at least one of them has a generic self?

Reply:
The query doesn't have to be fully concrete. In this specific case:
```
class B(T:! type) {}
impl forall [X:! type] B(X) as W {}
```
Then `B(X) as W` is true, and so `B(X) as L(bool)` does match 2nd impl, but does not match the type structure of `C(?) as L(bool)` of the first impl, which can be determined even though `B(X)` is not concrete.

Reply:
Rewrote.